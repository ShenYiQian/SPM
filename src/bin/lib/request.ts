import * as LibFs from "mz/fs";
import * as http from "http";
import * as qs from "querystring";
import {RequestMethod, SpmPackageRequest} from "./lib";

export function post(uri: string, params: Object, callback: (chunk: Buffer | string, resolve: any, reject: any) => void, handleResponse?: Function) {
    return new Promise((resolve, reject) => {
        // create request
        let reqOptions = SpmPackageRequest.getRequestOption(uri, RequestMethod.post);

        let req = http.request(reqOptions, (res) => {
            (handleResponse)
                ? handleResponse(res, resolve)
                : res.on('data', (chunk) => callback(chunk, resolve, reject));
        }).on('error', (e) => {
            reject(e);
        });

        // send content
        let reqParamsStr = qs.stringify(params);
        req.setHeader('Content-Type', 'application/x-www-form-urlencoded');
        req.setHeader('Content-Length', Buffer.byteLength(reqParamsStr, 'utf8').toString());
        req.write(reqParamsStr);
    });
}

export function postForm(uri: string, params: Object, filePath: Array<string>, callback: (chunk: Buffer | string, resolve: any, reject: any) => void) {
    return new Promise((resolve, reject) => {
        // create request
        let reqOptions = SpmPackageRequest.getRequestOption(uri, RequestMethod.post);
        let req = http.request(reqOptions, (res) => {
            res.on('data', (chunk) => callback(chunk, resolve, reject));
        }).on('error', (e) => {
            reject(e);
        });

        // send content
        let boundaryKey = Math.random().toString(16);
        let enddata = '\r\n----' + boundaryKey + '--';

        // build form params head
        let content = '';
        for (let key in params) {
            content += '\r\n----' + boundaryKey + '\r\n'
                + 'Content-Disposition: form-data; name="' + key + '" \r\n\r\n'
                + encodeURIComponent(params[key]);
        }

        let contentBinary = new Buffer(content, 'utf-8');
        let contentLength = contentBinary.length;

        // build upload file head
        if (filePath.length > 0) {
            content += '\r\n----' + boundaryKey + '\r\n'
                + 'Content-Type: application/octet-stream\r\n'
                + 'Content-Disposition: form-data; name="fileUpload"; filename="' + `${Math.random().toString(16)}.zip` + '"\r\n'
                + "Content-Transfer-Encoding: binary\r\n\r\n";
            contentBinary = new Buffer(content, 'utf-8');
            contentLength = LibFs.statSync(filePath[0]).size + contentBinary.length;

            // send request stream
            let fileStream = LibFs.createReadStream(filePath[0]);
            fileStream.on('end', () => {
                req.end(enddata);
            });
            fileStream.pipe(req, {end: false});
        }

        // send request headers
        req.setHeader('Content-Type', 'multipart/form-data; boundary=--' + boundaryKey);
        req.setHeader('Content-Length', `${contentLength + Buffer.byteLength(enddata)}`);
        req.write(contentBinary);

        if (filePath.length === 0) {
            req.end(enddata);
        }
    });
}