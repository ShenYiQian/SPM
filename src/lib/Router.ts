import * as Router from "koa-router";
import * as LibPath from 'path';
import * as LibFs from 'mz/fs';
import {ConfigOptions} from "./Config";
import {Connection} from "typeorm";

export interface ResponseSchema {
    code: number,
    msg: string | object
}
export type MiddlewareNext = () => Promise<any>

export default class RouteLoader {
    private static _instance: RouteLoader;

    private _initialized: boolean;
    private _router: Router;

    private constructor() {
        this._initialized = false;
        this._router = new Router();
    }

    public static instance(): RouteLoader {
        if (RouteLoader._instance === undefined) {
            RouteLoader._instance = new RouteLoader();
        }
        return RouteLoader._instance;
    }

    public async init(options: ConfigOptions, conn: Connection) {
        let _this = this;
        let dir = LibPath.join(__dirname, "api");
        let files = await LibFs.readdir(dir);

        for (let file of files) {
            if (LibPath.basename(file).match(/.+\.js$/) !== null) {
                await _this._createRouter(LibPath.join(dir, file), options, conn)
            }
        }

        this._initialized = true;
    }

    public getRouter(): Router {
        return this._router;
    }

    private async _createRouter(path: string, options: ConfigOptions, conn: Connection): Promise<void> {
        try {
            let api = require(path).api;
            this._router[api.method].apply(this._router, api.register(options, conn));
        } catch (err) {
            console.error(err.toString());
        }
    }
}