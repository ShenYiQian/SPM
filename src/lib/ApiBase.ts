import {Connection} from "typeorm";
import {ConfigOptions} from "./Config";
import {Context as KoaContext, Middleware as KoaMiddleware} from "koa";
import {MiddlewareNext, ResponseSchema} from "./Router";

export abstract class ApiBase {

    public method: string;
    public uri: string;
    public type: string;
    public options: ConfigOptions;
    public dbHandler: Connection;

    public abstract handle(ctx: KoaContext, next: MiddlewareNext): Promise<any>;

    public abstract paramsValidate(ctx: KoaContext);

    public register(options: ConfigOptions, dbHandler: Connection): Array<string | KoaMiddleware> {
        this.options = options;
        this.dbHandler = dbHandler;
        return [this.uri, this._validate(), this._execute()];
    };

    protected _validate(): KoaMiddleware {
        let _this = this;
        return async function (ctx: KoaContext, next: MiddlewareNext): Promise<void> {
            try {
                await _this.paramsValidate(ctx);
                await next();
            } catch (err) {
                ctx.body = _this.buildResponse(err.message, -1);
            }
        }
    }

    protected _execute(): KoaMiddleware {
        let _this: ApiBase = this;
        return async function (ctx: KoaContext, next: MiddlewareNext): Promise<void> {
            ctx.body = await _this.handle(ctx, next);
            await next();
        }
    }

    public buildResponse(msg: any, code: number = 0): ResponseSchema {
        if (code < 0) {
            console.log(`[${this.uri}]: ${msg}`)
        }

        return {
            code: code,
            msg: msg
        }
    }
}