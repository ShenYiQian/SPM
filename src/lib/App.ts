import * as Koa from "koa";
import * as koaBody from "koa-body";
// import * as koaBodyParser from "koa-bodyparser";
import Config from "./Config";
import Database from "./Database";
import Router from "./Router";

export default class App {
    private _initialized: boolean;
    public app: Koa;
    public config: Config;
    public database: Database;
    public router: Router;

    constructor() {
        this.app = new Koa();
        this.config = Config.instance();
        this.database = Database.instance();
        this.router = Router.instance();
        this._initialized = false;
    }

    public async init(): Promise<any> {
        await this.config.init();
        await this.database.init();
        await this.router.init(this.config.options, this.database.conn);

        this.app.use(koaBody({ multipart: true }));
        this.app.use(Router.instance().getRouter().routes());
        this._initialized = true;
    }

    public start(): void {
        if (!this._initialized) {
            return;
        }

        // server start
        this.app.listen(this.config.options.port, this.config.options.host, () => {
            console.log(`SPM Server start! ${this.config.options.host}:${this.config.options.port}`);
        });
    }
}