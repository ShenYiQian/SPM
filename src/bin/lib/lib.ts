import * as LibPath from "path";
import * as LibFs from "mz/fs";
import * as LibMkdirP from "mkdirp";
import * as bluebird from "bluebird";
import * as http from "http";
import * as recursive from "recursive-readdir";
import {ResponseSchema} from "../../lib/ApiBase";

export enum RequestMethod { post, get }

export interface SpmConfig {
    host: string;
    port: number;
    secret: string;
}

export interface SpmPackageConfig {
    name: string;
    version: string;
    description?: string;
    dependencies?: {
        [key: string]: string;
    };
}

export interface SpmPackage extends SpmPackageConfig {
    dependenciesChanged?: {
        [key: string]: string;
    };
    downloadUrl?: string;
    isDependencies?: boolean;
}

export interface SpmPackageMap {
    [dirname: string]: SpmPackage;
}

export const mkdir = bluebird.promisify<string>(LibMkdirP) as {
    <T>(path: string): bluebird<T>
};

export const rmdir = (dirPath: string) => {
    let files = LibFs.readdirSync(dirPath);
    if (files.length > 0) {
        for (let i = 0; i < files.length; i++) {
            let filePath = LibPath.join(dirPath, files[i]);
            if (LibFs.statSync(filePath).isFile()) {
                LibFs.unlinkSync(filePath);
            } else {
                rmdir(filePath);
            }
        }
    }
    LibFs.rmdirSync(dirPath);
};

export namespace Spm {

    export const INSTALL_DIR_NAME: string = 'spm_protos';
    export const SPM_VERSION_CONNECTOR: string = '__v';
    export const SPM_ROOT_PATH: string = LibPath.join(__dirname, '..', '..', '..');

    /**
     * Find project dir
     *
     * @returns {string}
     */
    export function getProjectDir(): string {
        return process.cwd();
    }

    /**
     * Save secret value into .spmlrc
     *
     * @returns {void}
     */
    export function saveSecret(value: string) {
        let lrcPath = LibPath.join(Spm.getProjectDir(), '.spmlrc');
        LibFs.writeFileSync(lrcPath, value, 'utf-8');
    }

    /**
     * Load secret value from .spmlrc
     *
     * @returns {string}
     */
    export function loadSecret(): string {
        let lrcPath = LibPath.join(Spm.getProjectDir(), '.spmlrc');
        if (LibFs.existsSync(lrcPath) && LibFs.statSync(lrcPath).isFile()) {
            return LibFs.readFileSync(lrcPath).toString();
        } else {
            return '';
        }
    }

    /**
     * Get Spm config
     *
     * @returns {SpmConfig}
     */
    export function getConfig(path?: string): SpmConfig {
        let configPath = (path) ? path : LibPath.join(SPM_ROOT_PATH, 'config', 'config.json');
        if (LibFs.existsSync(configPath) && LibFs.statSync(configPath).isFile()) {
            return JSON.parse(LibFs.readFileSync(configPath).toString());
        } else {
            throw new Error('[Config] config file path have to be an absolute path!');
        }
    }

    /**
     * Read spm.json via config path
     *
     * @param {string} path
     * @returns {SpmPackageConfig}
     */
    export function getSpmPackageConfig(path: string): SpmPackageConfig {
        return JSON.parse(LibFs.readFileSync(path).toString());
    }

    /**
     * Get all installed SpmPackage from projectDir
     *
     * @returns {Promise<SpmPackageMap>}
     */
    export async function getInstalledSpmPackageMap(path?: string): Promise<SpmPackageMap> {
        let projectDir = path ? path : Spm.getProjectDir();
        let installDir = LibPath.join(projectDir, Spm.INSTALL_DIR_NAME);

        if (LibFs.existsSync(installDir) && LibFs.statSync(installDir).isDirectory()) {
            let files = await recursive(installDir, ['.DS_Store']);

            let spmPackageMap = {} as SpmPackageMap;
            for (let file of files) {
                let basename = LibPath.basename(file);
                if (basename.match(/.+\.json/) !== null) {
                    let dirname = LibPath.dirname(file).replace(installDir, '').replace('\\', '').replace('/', '');
                    let packageConfig = Spm.getSpmPackageConfig(file);
                    spmPackageMap[dirname] = {
                        name: packageConfig.name,
                        version: packageConfig.version,
                        dependencies: packageConfig.dependencies
                    };
                }
            }
            return spmPackageMap;
        } else {
            return {};
        }
    }

    /**
     * Replace string in file
     *
     * @param {string} filePath
     * @param {Array<[RegExp, any]>} conditions
     * @returns {Promise<void>}
     */
    export async function replaceStringInFile(filePath: string, conditions: Array<[RegExp, any]>) {
        try {
            let buffer = await LibFs.readFile(filePath);
            let content = buffer.toString();
            for (let [reg, word] of conditions) {
                content = content.replace(reg, word);
            }
            LibFs.writeFileSync(filePath, Buffer.from(content));
        } catch (e) {
            throw e;
        }
    }
}

export namespace SpmPackageRequest {
    /**
     * Get Spm Publish request config
     *
     * @returns {SpmConfig}
     */
    export function getRequestOption(path: string, method: RequestMethod = RequestMethod.get): http.RequestOptions {
        let spmHttpConfig = Spm.getConfig();
        return {
            host: spmHttpConfig.host,
            port: spmHttpConfig.port,
            method: RequestMethod[method],
            path: path,
        };
    }

    /**
     * Parse request response
     *
     * @param {string} chunk
     * @returns {any}
     */
    export function parseResponse(chunk: Buffer | string): any {
        let response = JSON.parse(chunk.toString()) as ResponseSchema;
        if (response.code < 0) {
            throw new Error(response.msg.toString());
        }
        return response.msg;
    }
}