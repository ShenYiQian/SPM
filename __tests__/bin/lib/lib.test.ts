import * as LibFs from 'mz/fs';
import * as LibPath from 'path';
import * as lib from "../../../src/bin/lib/lib";

// CONST
const TMP_DIR_NAME = "tmp";

describe('mkdir', async () => {

    beforeAll(() => {
        process.chdir('demo');
    });

    afterAll(async() => {
        await lib.rmdir(LibPath.join(process.cwd(), TMP_DIR_NAME));
        process.chdir('..');
    });

    // test
    test('test mkdir is succeed', async() => {
        // run
        await lib.mkdir(LibPath.join(process.cwd(), TMP_DIR_NAME));

        // test
        let files = await LibFs.readdir(process.cwd());
        expect(files[files.indexOf(TMP_DIR_NAME)]).toEqual(TMP_DIR_NAME);
    });

});

describe('rmdir', async () => {

    beforeAll(async() => {
        process.chdir('demo');
        await lib.mkdir(LibPath.join(process.cwd(), TMP_DIR_NAME));
        await lib.mkdir(LibPath.join(process.cwd(), TMP_DIR_NAME, "mock"));
        await LibFs.writeFile(LibPath.join(process.cwd(), TMP_DIR_NAME, "mock.txt"), "mock");
    });

    afterAll(async() => {
        process.chdir('..');
    });

    // test
    test('test rmdir is succeed', async() => {
        // run
        await lib.rmdir(LibPath.join(process.cwd(), TMP_DIR_NAME));

        // test
        let files = await LibFs.readdir(process.cwd());
        expect(files.indexOf(TMP_DIR_NAME)).toEqual(-1);
    });

});

describe('Spm', async () => {
    let lrcPath = LibPath.join(process.cwd(), 'demo', '.spmlrc');

    beforeAll(() => {
        process.chdir('demo');
        if (LibFs.existsSync(lrcPath)) {
            LibFs.unlinkSync(lrcPath);
        }
    });

    afterAll(() => {
        process.chdir('..');
    });

    test('getProjectDir', () => {
        expect(lib.Spm.getProjectDir()).toEqual(process.cwd());
    });

    test('SaveSecret/LoadSecret: secret is saved', () => {
        lib.Spm.saveSecret('1q2w3e4r');
        expect(lib.Spm.loadSecret()).toEqual('1q2w3e4r');
    });

    test('LoadSecret: unlink .spmlrc file will return ""', () => {
        // run
        if (LibFs.existsSync(lrcPath)) {
            LibFs.unlinkSync(lrcPath);
        }
        // test
        expect(lib.Spm.loadSecret()).toEqual('');
    });

    test('GetConfig: toMatchSnapshot()', () => {
        expect(lib.Spm.getConfig()).toMatchSnapshot();
    });

    test('GetConfig: toThrow(\'[Config] config file path have to be an absolute path!\')', () => {
        expect(() => {
            lib.Spm.getConfig(LibPath.join(process.cwd(), 'config.json'))
        }).toThrow('[Config] config file path have to be an absolute path!');
    });

    test('GetSpmPackageConfig: toMatchSnapshot()',() => {
        expect(lib.Spm.getSpmPackageConfig(LibPath.join(process.cwd(), 'spm.json'))).toMatchSnapshot();
    });

    test('GetInstalledSpmPackageMap: toMatchSnapshot()', async () => {
        await expect(lib.Spm.getInstalledSpmPackageMap()).resolves.toMatchSnapshot();
        await expect(lib.Spm.getInstalledSpmPackageMap(LibPath.join(process.cwd(), '..'))).resolves.toEqual({});
    });

    test('ReplaceStringInFile: toEqual(1.0.0)', async () => {
        const configPath = LibPath.join(process.cwd(), 'spm.json');
        await lib.Spm.replaceStringInFile(configPath, [
            [new RegExp(`0.0.0`, 'g'), `1.0.0`]
        ]);

        expect(lib.Spm.getSpmPackageConfig(configPath).version).toEqual('1.0.0');

        await lib.Spm.replaceStringInFile(configPath, [
            [new RegExp(`1.0.0`, 'g'), `0.0.0`]
        ]);
    });

    test('ReplaceStringInFile: toThrow()', () => {
        const configPath = LibPath.join(process.cwd(), 'spm_error.json');
        expect(() => { lib.Spm.getSpmPackageConfig(configPath) }).toThrowErrorMatchingSnapshot();
    });
});

describe('SpmPackageRequest', async () => {
    const emoji = "♈ ♉ ♊ ♋ ♌ ♍ ♎ ♏ ♐ ♑ ♒ ♓";

    test('GetRequestOption: toMatchSnapshot()', () => {
        expect(lib.SpmPackageRequest.getRequestOption('/mock', lib.RequestMethod.post)).toMatchSnapshot();
    });

    test('ParseResponse: toEqual(♈ ♉ ♊ ♋ ♌ ♍ ♎ ♏ ♐ ♑ ♒ ♓)', () => {
        expect(lib.SpmPackageRequest.parseResponse(JSON.stringify({code: 0, msg: emoji}))).toEqual(emoji);
    });

    test('ParseResponse: toThrow("something error")', () => {
        expect(() => {
            lib.SpmPackageRequest.parseResponse(JSON.stringify({code: -1, msg: "something error"}))
        }).toThrow("something error");
    });
});