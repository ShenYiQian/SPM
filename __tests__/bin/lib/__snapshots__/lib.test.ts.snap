// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spm GetConfig: toMatchSnapshot() 1`] = `
Object {
  "host": "172.17.94.239",
  "port": 3000,
  "secret": "1q2w3e4r",
}
`;

exports[`Spm GetInstalledSpmPackageMap: toMatchSnapshot() 1`] = `Object {}`;

exports[`Spm GetSpmPackageConfig: toMatchSnapshot() 1`] = `
Object {
  "dependencies": Object {
    "bookPb": "0.0.0",
    "userPb": "2.0.0",
  },
  "name": "demo",
  "version": "0.0.0",
}
`;

exports[`Spm ReplaceStringInFile: toThrow() 1`] = `"ENOENT: no such file or directory, open 'E:\\\\data1\\\\www\\\\demo\\\\SASDN-PM\\\\demo\\\\spm_error.json'"`;

exports[`SpmPackageRequest GetRequestOption: toMatchSnapshot() 1`] = `
Object {
  "host": "172.17.94.239",
  "method": "post",
  "path": "/mock",
  "port": 3000,
}
`;
